cranelift/isle/isle/src/trie.rs

```diff
@@ -196,28 +219,76 @@ impl TrieNode {
         if self.is_empty() {
-            *self = TrieNode::Decision { edges: vec![] };
+            *self = TrieNode::Decision {
+                edges: vec![],
+                last_prio: None,
+                cur_prio: None,
+            };
         }
 
         // We must be a decision node.
-        let edges = match self {
-            &mut TrieNode::Decision { ref mut edges } => edges,
+        let (edges, last_prio, cur_prio) = match self {
+            &mut TrieNode::Decision {
+                ref mut edges,
+                ref mut last_prio,
+                ref mut cur_prio,
+            } => (edges, last_prio, cur_prio),
             _ => panic!("insert on leaf node!"),
         };
 
+        // If we are inserting at a lower prio than in `cur_prio`,
+        // `cur_prio` moves to `last_prio` (and controls our minimum
+        // insertion point) and we initialize `cur_prio` with our
+        // current priority and new max insertion index (of 0).
+        if cur_prio.is_none() || prio < cur_prio.unwrap().prio {
+            *last_prio = *cur_prio;
+            *cur_prio = Some(PrioFrontier {
+                prio,
+                // This is `None` initially but will be updated below
+                // to at least the frontier, as we always reuse or
+                // insert an edge and that edge is always >=
+                // `last_prio.edge_idx`, if not `None`.
+                edge_idx: None,
+            });
+        }
+        let cur_prio = cur_prio.as_mut().unwrap();
+
+        // Determine the minimum edge index under which we can insert
+        // while respecting priorities.
+        let start = last_prio.and_then(|frontier| frontier.edge_idx);
+
         // Now find or insert the appropriate edge.
         let edge = edges
             .iter()
-            .position(|edge| edge.symbol == op && edge.prio == prio)
+            .skip(start.unwrap_or(0))
+            .position(|edge| edge.symbol == op)
+            .map(|pos| pos + start.unwrap_or(0))
             .unwrap_or_else(|| {
-                edges.push(TrieEdge {
-                    prio,
-                    symbol: op,
-                    node: TrieNode::Empty,
-                });
-                edges.len() - 1
+                // Insert in a position among our allowed range
+                // (strictly after `start` now) that would be sorted
+                // according to the `op` symbol.
+                let first_after_prev_prio = start.map(|x| x + 1).unwrap_or(0);
+                assert!(first_after_prev_prio <= edges.len());
+                let insert_pos = edges[first_after_prev_prio..]
+                    .binary_search_by(|edge| edge.symbol.cmp(&op))
+                    .unwrap_err()
+                    + first_after_prev_prio;
+
+                if cur_prio.edge_idx.is_some() && insert_pos <= cur_prio.edge_idx.unwrap() {
+                    *cur_prio.edge_idx.as_mut().unwrap() += 1;
+                }
```

Comment:
Ah right, inserting between `start` and `cur-prio.edge_idx` means that `cur_prio.edge_idx` needs to be incremented to point at the same element :+1:

Reply:
This is also the case that currently requires that `cur_prio.edge_idx` be set to `None` on line 251. Is it worth calling that connection out with a comment?

Reply:
Yup, added (actually at the comment above this where `None` is set).

Reply:
Yep, I was hoping to add another comment here referencing that one to tie the two together.

Reply:
Ah, sure thing -- added one here too!